feat(all): expand

martsokha opened the pull request:
Changes so far:

- add features for using both `reqwest/native-tls` (enabled by default) and `reqwest/rustls-tls`
- replace feature `async` with `blocking`, same as in `reqwest`
- add crate-specific `Result` type alias
- service-based client, similar to https://github.com/resend/resend-go
- add `base_url` support with an env var `RESEND_BASE_URL`
- add `impl Default` for a client with an env var `RESEND_API_KEY`
- add `USER-AGENT` to the request with `RESEND_USER_AGENT`

AntoniosBarotsis commented:
This is *quite* a lot of changes 😅. The PR has multiple times more effort put into it than the entire project which is interesting.

I have been taking a few looks at this in the last few days  even though there's a decent amount of `todo`s and stuff. Do write any messages in here if you want to discuss anything, I'll be sure to reply asap.

AntoniosBarotsis commented on README.md:
```diff
@@ -3,11 +3,33 @@
-Emails are sent via the [`resend_client::ResendClient`] which provides both a synchronous and
-asynchronous send method. The two are mutually exclusive and accessible via the `async` feature. The
-crate uses [`reqwest`](https://github.com/seanmonstar/reqwest) internally.
+Emails are sent via the [`Resend`] client which provides both a
```
Just a small note here, I had "[`resend_client::ResendClient`]" before because I use the readme in the [docs](https://docs.rs/resend-rs/latest/resend_rs/) and it links to the struct properly. More noting this down for myself if anything, it is probably better we move/copy the code examples in the docs directly instead of the readme.

martsokha replied:
Yeah, I've reexported the `Client` into the lib so that docs should look fine now. 
Agree, I'll move code examples into the `lib.rs`.

The main issue right now is that some of the types generated with `openapi-generator -i` [resend.yaml][yaml] appear to be either outdated (according to the [API reference][API]) (e.g., available region list) or not entirely correct (e.g. options on required params).

[yaml]: https://raw.githubusercontent.com/resend/resend-openapi/main/resend.yaml
[API]: https://resend.com/docs/api-reference

AntoniosBarotsis commented on src/api_keys.rs:
```diff
@@ -0,0 +1,192 @@
+    pub async fn create(&self, api_key: CreateApiKeyRequest) -> Result<CreateApiKeyResponse> {
+        let request = self.0.build(Method::POST, "/api-keys");
+        let response = request.json(&api_key).send().await?;
+        let content = response.json::<CreateApiKeyResponse>().await?;
```
The issue with this is that it tries to parse the API response into the `CreateApiKeyResponse` struct but when the server returns an error, it tries to parse that instead and thus the API error is both not reported and you get a confusing parsing error instead. 

You could test this by for example sending an email to an address that is not your resend address if you are still in testing as their docs state "You can only send testing emails to your own email address".

I think the way to go about this would be to represent all [resend errors](https://resend.com/docs/api-reference/errors). They did a great job at writing them down in that page, I think we could show the same information in an error message (or direct the users to the page itself by giving them a link such as `https://resend.com/docs/api-reference/errors#invalid-from-address`).

We are already returning a Result from all of these methods so it would not be too hard to add a `ResendError` entry to the enum and have a "is status 200" before parsing it here so we can return the errors properly. The same goes for all API calls.

martsokha replied:
Oops, forgot about `.error_for_status()?`.
Yeah, I thought about getting rid of `reqwest::Error` and using actual error types.
Let me try to figure something out.

martsokha replied:
Updated `crate::Error` and `emails.rs::send()`. Should I make any changes before updating the rest?

AntoniosBarotsis replied:
It's great now!

martsokha replied:
Updated again to use the `ErrorKind`.

AntoniosBarotsis commented on src/lib.rs:
````diff
@@ -1,5 +1,63 @@
+#![forbid(unsafe_code)]
+#![cfg_attr(docsrs, feature(doc_cfg))]
 #![doc = include_str!("../README.md")]
+//! #### Examples
+//!
+//! ```rust
+//! use resend_rs::{Client, Result};
+//!
+//! #[tokio::main]
+//! async fn main() -> Result<()> {
+//! //     let resend = Resend::default();
+//! //     let _ = resend.emails.send().await?;
+//!
+//!     Ok(())
+//! }
+//! ```
````
This is slightly outdated, we could use something like this:

```
#![forbid(unsafe_code)]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![doc = include_str!("../README.md")]
//! #### Examples
//!
//! ```rust,no_run
//! use resend_rs::{types::SendEmailRequest, Client, Result};
//! 
//! #[tokio::main]
//! async fn main() -> Result<()> {
//!     let resend = Client::default();
//! 
//!     let req = SendEmailRequest::new(
//!         "rss2email@resend.dev".to_string(),
//!         vec!["antonis.barotsis@gmail.com".to_string()],
//!         "subject".to_string(),
//!     )
//!     .with_text("Hello world!");
//! 
//!     let res = resend.emails.send(req).await?;
//! 
//!     println!("{}", res.id);
//! 
//!     Ok(())
//! }
//! ```
```

Feel free to make any changes to this. It might also be a good idea to make this change at the very end in case we make any relevant changes in the meantime.

Edit: I just now realized I almost rewrote the test you made in [`email.rs`](https://github.com/martsokha/resend-rs/blob/6e7b2319bd37a25b71b55d9e2165a7e546e3b5d7/src/emails.rs#L351-L371), feel free to use that instead, just remember the `no_run`.

martsokha replied:
Removed test from `email.rs`, moved it here with `no_run`.

AntoniosBarotsis commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        let path = self
+            .base_url
+            .join(path)
+            .expect("should a valid API endpoint");
```
"should *be* a valid API endpoint" :)

martsokha replied:
Didn't notice, fixed.

AntoniosBarotsis commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+    /// Panics if the environment variable `RESEND_BASE_URL` is present but is not a valid URL.
+    ///
+    /// [`Resend`]: https://resend.com
+    /// [`reqwest::Client`]: ReqwestClient
```
This line seems to actually break the doc comment, removing it fixes the link, not sure if you meant to have these the other way around maybe

martsokha replied:
What line, exactly? Both links seem to work for me.
Here is what I see in IDEA and cargo doc:
https://imgur.com/a/HGCMkO1

AntoniosBarotsis replied:
Ok today I learned vscode does this wrong lol, if I built the docs, it works as it is now and breaks if you remove the last line. No clue why vscode apparently does its own thing here.

AntoniosBarotsis commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    /// Returns the used `User-Agent` header value.
+    #[must_use]
```
What is the rationale behind these? I've never used them myself. I know they raise a warning when you don't use the returned value but why add it in these methods specifically? Asking out of curiosity, not really a review comment 😅

martsokha replied:
Because these methods don't have any side effects, not using the return value is likely to result in a logical error. The same applies to responses from retrieve and list endpoints.

AntoniosBarotsis replied:
Makes sense, thanks!

AntoniosBarotsis commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+    ///
+    /// ### Panics
+    ///
+    /// Panics if the environment variable `RESEND_API_KEY` is not a valid API key.
```
Bit of a nitpick, this panics if the API key is *not set*, not necessarily invalid.

martsokha replied:
Fixed.

AntoniosBarotsis reviewed (changes requested):
Most looks great! Just a reminder, you've left some `todo`s for yourself in `lib.rs`.

martsokha commented on Cargo.toml:
```diff
@@ -1,16 +1,37 @@
+keywords = ["email", "resend"]
+description = "A minimal Resend client."
+
+[package.metadata.docs.rs]
+all-features = false
+rustdoc-args = ["--cfg", "docsrs"]
```
I originally wanted to make the blocking methods not mutually exclusive. 
I'm not sure whether to use a prefix or suffix, or to implement a separate client.

AntoniosBarotsis replied:
I'm just confused what those `#[cfg_attr(docsrs, doc(cfg(not(feature = "blocking"))))]` and `#[cfg_attr(docsrs, doc(cfg(feature = "blocking")))]` lines do

martsokha replied:
Oh, they add a little hint `Available on crate feature blocking only`, e.g.
- on modules: https://docs.rs/tokio/latest/tokio/index.html
- on method `id`: https://docs.rs/tokio/latest/tokio/task/struct.AbortHandle.html

AntoniosBarotsis replied:
Ok so I was confused why I was not getting that when building the docs locally, [apparently you need to set `[RUSTDOCFLAGS="--cfg docsrs"`](https://users.rust-lang.org/t/how-to-test-a-package-metadata-docs-rs-setting/53818). Building with `cargo doc --no-deps` does not show all methods though

![image](https://github.com/AntoniosBarotsis/resend-rs/assets/50240570/e1b6c730-f268-4a99-81bc-35b77883555e)

If I understand correctly this should have each method essentially duplicated, one time for the blocking and one for the async version right?

martsokha replied:
Yes, async with feature blocking disabled, and vice versa. 
The easy fix would be to prefix all blocking methods with 'blocking_' and then remove 'cfg_attr(docsrs)' from all non-blocking methods.

AntoniosBarotsis replied:
I think no one has really figured out a good way of dealing with this sync+async sitation, it is just one of the few shortcomings of the language. I think it is fine as is, if you have nothing else to add to this thread feel free to resolve it 👍

martsokha replied:
I honestly don't like them to have the same name as the non-blocking ones since they are not visible in the docs unless you enable all features.

AntoniosBarotsis replied:
Yeah that's fair. Maybe we can keep the async ones as they are right now and postfix the sync ones with `_sync`?

martsokha replied:
So, I have found this `maybe_async` crate. 
It's definitely not the optimal solution, but it's better than just copy-pasting every single handler.

AntoniosBarotsis commented on CHANGELOG.md:
```diff
@@ -1,17 +1,19 @@
+and this project adheres to
+[Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 <!-- ## [Unreleased] -->
```
Would be nice to have a before-after conversion similar to what I had in v0.3. Since there was just 1 method before this PR this should be easy. I have a code example [here](https://github.com/AntoniosBarotsis/resend-rs/blob/88f3d18b3c471b87f5b8d3281c948e82ae21a9d0/src/resend_client.rs#L13-L26).

martsokha replied:
The weird formatting is a result of me using deno fmt for markdown files.
I'm a bit confused, do you want an example in the client documentation?

AntoniosBarotsis replied:
I think Github likes showing 6 lines minimum because I specifically selected the "unreleased" line. Meant to add an example similar to this: 
![image](https://github.com/AntoniosBarotsis/resend-rs/assets/50240570/ad947841-c758-45c1-8fd2-afb8d54d7e67)

martsokha commented on README.md:
```diff
@@ -1,13 +1,37 @@
+- `RESEND_API_KEY` to enable `impl Default` for a `Client`.
+- `RESEND_BASE_URL` to override the default base address:
+  `https://api.resend.com`.
+- `RESEND_USER_AGENT` to override the default user-agent: `resend-rs/0.1.0`.
```
I'm not sure if I did it correctly. 🤔

AntoniosBarotsis replied:
Looks correct, I would just write it something like 

"`RESEND_API_KEY` to enable `impl Default` for a `Client`. (Required)" just to make it stupidly obvious you need to set this. I know my eyes skip over stuff like this all the time that's why I suggested this haha

martsokha commented:
- Should we `impl Deref` for a `Client` to `ClientInner` (same as the `Client` right now) to limit cloning? Could be useful for something like a `State` extractor in `Axum`. 
- What should we do with `Option`s in responses, for example, `CreateAudienceResponse`?

AntoniosBarotsis commented:
- Don't see why not
- I should've read the API docs more carefully cause I'm guessing in most cases the responses will always have all the fields unless specified otherwise in the docs so we should probably remove them where it makes sense. Were the options added initially because we were not handling error responses?

martsokha commented:
> - Don't see why not
> - I should've read the API docs more carefully cause I'm guessing in most cases the responses will always have all the fields unless specified otherwise in the docs so we should probably remove them where it makes sense. Were the options added initially because we were not handling error responses?

No, options were automatically generated by the openapi-generator from the .yaml file.
https://resend.com/changelog/introducing-resends-openapi-spec

AntoniosBarotsis commented:
Ah makes sense. I'm not 100% sure on this but from the docs it does not look like these should be optional which is weird that they then get generated like this. I think it seems safe to remove them but I wonder if they get generated like that with the idea that a response might have a completely different structure if it was an error (thus each param is optional cause it might not exist, to avoid crashes). What do you think?

martsokha commented:
Looks like the response fields in `resend-go` are tagged with `json:"omitempty"`.
I suppose I'll leave them optional in this case.

https://github.com/resend/resend-go/blob/main/emails.go#L12

AntoniosBarotsis commented:
Fair enough

martsokha commented:
- remove `object` fields from response types.
- remove `*DataResponse` type (used in the `List*Response`), replaced with a
  `Retrieve*Response`.
- hide duplicated response type data, e.g. `RemoveAudienceResponse` returns both
  `id` and `deleted`, but the `id` is the input of a `Audience::delete`. Changed
  `Audience::delete` to return `deleted` only.

AntoniosBarotsis commented:
Made 2 small changes (remember to pull!);

- The doctest was giving me a compile error so I fixed that, the example should look the same as before in the docs
- I was getting rate limited and thus 1 random test was failing most of the time. For now I just made an alias to run the tests on one thread but I'll probably have to do something about this in the future.

I think everything else is fine from me for now, there's just a couple `TODO`s you've left yourself .

*Also I really feel like the tests should run fine in CI but they are refusing to, I'll give it a few more attempts and if they still fail I'll just fix them after this is merged*

AntoniosBarotsis commented:
What I think might be happening is that Github might not be setting the RESEND_API_KEY variable because it's a secret and this is a PR. I get the exact same error message if I run the tests locally with `RESEND_API_KEY=''` so I'm 99% the tests are actually fine.

AntoniosBarotsis commented:
[That indeed seems to be the case](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/);

> Due to the dangers inherent to automatic processing of PRs, GitHub’s standard `pull_request` workflow trigger by default prevents write permissions and secrets access to the target repository. However, in some scenarios such access is needed to properly process the PR. To this end the `pull_request_target` workflow trigger was introduced.

I was indeed skeptical at first about needing to add this API key for the action but considering it is inaccessible by default it is not that bad actually so I will keep it like this for now.

martsokha commented:
- added `*Id` types for a bit more type-safety after using `Contact`/`Audience` ids in the wrong order.

I believe it's possible to wrap a `reqwest::Client` with a `tower::limit::RateLimit` middleware to solve rate limit issues.

martsokha commented:
I'm not sure if we really need custom `Serialize` and `Deserialize` for `Region`.

AntoniosBarotsis commented on src/api_keys.rs:
```diff
@@ -139,21 +166,27 @@ pub mod types {
         pub created_at: String,
+        #[cfg(feature = "time")]
+        #[serde(with = "time::serde::iso8601")]
+        pub created_at: time::OffsetDateTime,
```
This is up for discussion but I feel like having these be *not* mutually exclusive would be better here. [The official docs don't really like mutually exclusive features](https://doc.rust-lang.org/cargo/reference/features.html#mutually-exclusive-features) and I think for most stuff other than sync/async splits they are not worth it. The crate's code becomes a bit messier but more importantly, if someone is using this crate without the `time` feature initially and they later on realize that it would be neat to have the relevant types, it is possible that some of their code now fails to compile if they used the fact that the field was a string before.

martsokha replied:
what name should it use? `created_at: String` and `created_at_time: time::OffsetDateTime`?

AntoniosBarotsis replied:
Yeah I also tried thinking of what to name it but didn't come up with something non-ugly. Maybe we could do `created_at_datetime`? That at least feels like a type. Not sure if this needs custom (de)serialize traits, I've always used chrono instead of time but I would assume time has methods for parsing dates (?)

martsokha replied:
Yeah, time has these methods, but then we'd need to `#[serde(skip)] created_at` and manually parse the `created_at_time` after calling `response.json()`.

AntoniosBarotsis replied:
Ah I see what you mean. [Doesn't the `serde` feature make this work though](https://time-rs.github.io/book/how-to/parse-dates.html#parsing-into-structs-with-serde)?

martsokha replied:
it does work, but the "created_at" field consumes that part of the response json, and afaik it's not possible to parse the same value twice without impl custom serialize/deserialize traits. at least it used to be this way.

AntoniosBarotsis replied:
Oh dang I see., that's annoying. I think implementing deserialize manually would be the most convenient for anyone using this crate. We could also put the parsed time behind a function instead of a field but that would be weird since everything else is a field (+ that would need to be parsed on each function call, unless we used `OnceLock` or something similar).

AntoniosBarotsis replied:
Actually it might be possible to use [this](https://serde.rs/container-attrs.html#from)

martsokha replied:
or we could just get rid of a feature and make it `OffsetDateTime`  by default

AntoniosBarotsis replied:
Would something like this work?

```rs
    #[must_use]
    #[derive(Debug, Clone, Deserialize)]
    #[serde(from = "ApiKeyRaw")]
    pub struct ApiKey {
        /// The ID of the API key.
        pub id: ApiKeyId,
        /// The name of the API key.
        pub name: String,

        /// The date and time the API key was created.
        pub created_at: String,
        #[cfg(feature = "time")]
        pub created_at_time: time::OffsetDateTime,
    }

    impl From<ApiKeyRaw> for ApiKey {
        fn from(value: ApiKeyRaw) -> Self {
            todo!()
        }
    }

    #[derive(Debug, Clone, Deserialize)]
    struct ApiKeyRaw {
        pub id: ApiKeyId,
        pub name: String,
        pub created_at: String,
    }
```

AntoniosBarotsis replied:
Quick and dirty `From` implementation

```rs
    impl From<ApiKeyRaw> for ApiKey {
        fn from(value: ApiKeyRaw) -> Self {
            ApiKey { 
                id: value.id, 
                name: value.name, 
                created_at: value.created_at.clone(), 
                created_at_time: time::OffsetDateTime::parse(&value.created_at, &Iso8601::DEFAULT).unwrap()
            }
        }
    }
```

AntoniosBarotsis replied:
The test in that file passes for me so I'm guessing it works 😅 

Let me know what you think

martsokha replied:
I think so, but that would be Email, Audience, Contact, Domain, ApiKey and DomainReply (idk about this one) duplicates.
The only reason I'd like to have a parsed datetime is to make it easier to navigate logs. Having the date repeated doesn't really add much value in this context.

martsokha replied:
Btw, should we remove DomainReply? The only difference seems to be the 
```rust
#[serde(rename = "dnsProvider")]
pub dns_provider: String,
```
field, it could be replaced it with `Domain` with additional `pub dns_provider: String,` that would be present only as a response of a `Domain::create`.

AntoniosBarotsis replied:
> I think so, but that would be Email, Audience, Contact, Domain, ApiKey and DomainReply (idk about this one) duplicates.

Well the `ApiKeyRaw` would be just an internal type not exposed anywhere so for the users this wouldn't be tedious. For this crate I mean it's a bit annoying but it is not that bad.

AntoniosBarotsis replied:
For the domain reply, I think it is fine now, especially since all our reply types have `Reply` in the type name so it would be confusing for this to be an exception

martsokha replied:
I think it's the only `*Reply` left 😄

AntoniosBarotsis replied:
I've lost track of some of the changes at this point 🤣 makes sense then haha

martsokha replied:
Yeah, I've renamed most of the public `Method*Request` and `Method*Response` classes. Some were reduced to a single argument (e.g. `CreateAudienceRequest`), while others were incorrectly generated by an open-api generator (e.g. `ListAudiencesItemResponse`, which was supposed to be `GetAudienceResponse`).

AntoniosBarotsis replied:
Any thoughts on the time parsing?

martsokha replied:
Not a big fan of seeing the same thing repeated twice (both `created_at*` and `*Raw`), but I'm fine with that.

AntoniosBarotsis replied:
Do you think it would be better to manually implement deserialize instead?

martsokha replied:
We'd still have `created_at`/`created_at_time`. Pretty much the same thing (if you print it). 
And it's the same ~200 loc, just more complicated.

AntoniosBarotsis replied:
Ah I see what you mean. I still feel like having that duplication is better than mutually excluding them tho

martsokha replied:
Agree, that's why I'd rather avoid both issues by either removing time support or by making it the default.

AntoniosBarotsis replied:
I think removing it all together would be reasonable as well, considering we are providing the string date anyway, anyone that wants to parse it can use whichever crate they want to do so. Maybe that's the best solution to be honest

martsokha replied:
Done

AntoniosBarotsis commented on src/audiences.rs:
```diff
@@ -116,24 +123,29 @@ pub mod types {
         pub created_at: String,
+        #[cfg(feature = "time")]
+        #[serde(with = "time::serde::iso8601")]
+        pub created_at: time::OffsetDateTime,
```
Same here

AntoniosBarotsis commented on src/contacts.rs:
```diff
@@ -202,15 +204,21 @@ pub mod types {
-        pub created_at: String,
         /// Indicates if the contact is unsubscribed.
         pub unsubscribed: bool,
+
+        /// Timestamp indicating when the contact was created.
+        #[cfg(not(feature = "time"))]
+        pub created_at: String,
+        #[cfg(feature = "time")]
+        #[serde(with = "time::serde::iso8601")]
+        pub created_at: time::OffsetDateTime,
```
Same here

AntoniosBarotsis commented on src/domains.rs:
```diff
@@ -124,53 +131,144 @@ pub mod types {
+
+            deserializer.deserialize_str(RegionVisitor)
+        }
     }
```
For this wouldn't it possible to use `serde(rename)`s instead of implementing serializing and deserializing? It's been a while since I've used serde but wouldn't this work:

```rs
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Region {
    /// 'us-east-1'
    #[serde(rename = "us-east-1")]
    UsEast1,
    /// 'eu-west-1'
    #[serde(rename = "eu-west-1")]
    EuWest1,
    /// 'sa-east-1'
    #[serde(rename = "sa-east-1")]
    SaEast1,
    /// 'ap-northeast-1'
    #[serde(rename = "ap-northeast-1")]
    ApNorthEast1,
}
```

Also judging from the docs, I don't think there's a reason to have the `Custom(String)` entry, is this something the Go library does? Because [the docs](https://resend.com/docs/api-reference/domains/create-domain) specifically mention:

> Possible values: `us-east-1' | 'eu-west-1' | 'sa-east-1' | 'ap-northeast-1'`

martsokha replied:
That's what It used to be, but how do you add `Custom(String)`?

AntoniosBarotsis replied:
Considering that there's only 4 regions I think we can remove the `Custom` entry, it is trivial to add more regions to the enum when the API adds them

martsokha replied:
Alright, reverted and added #[non_exhaustive].

AntoniosBarotsis commented on src/domains.rs:
```diff
@@ -124,53 +131,144 @@ pub mod types {
         pub status: String,
+
+        /// The date and time the domain was created.
+        #[cfg(not(feature = "time"))]
+        pub created_at: String,
+        #[cfg(feature = "time")]
+        #[serde(with = "time::serde::iso8601")]
+        pub created_at: time::OffsetDateTime,
```
Same here

AntoniosBarotsis commented on src/domains.rs:
```diff
@@ -191,31 +289,40 @@ pub mod types {
         pub created_at: String,
+        #[cfg(feature = "time")]
+        #[serde(with = "time::serde::iso8601")]
+        pub created_at: time::OffsetDateTime,
```
Same here

AntoniosBarotsis commented on src/emails.rs:
```diff
@@ -356,8 +357,13 @@ pub mod types {
         pub created_at: Option<String>,
+        #[cfg(feature = "time")]
+        #[serde(with = "time::serde::iso8601")]
+        pub created_at: Option<time::OffsetDateTime>,
```
Same here

AntoniosBarotsis reviewed (commented):
Left the same comment like 6 times just so we don't miss any of the usages 😅

AntoniosBarotsis commented:
What is there left to do at this point?

martsokha commented:
- urlencode (not sure if actually needed)
- rate limit
- domain.rs tests and SendEmain::new 
- type renames if any needed

martsokha commented:
the only issue with rate limit is to make it work with blocking, otherwise it seems to be pretty easy

martsokha commented:
- Not sure what to do about the rate limit for a blocking client, should we just ignore it or maybe `sleep` for `1/reqs_per_sec seconds`? Meh.
- Can you check out those last 3 TODOs?

AntoniosBarotsis commented:
I think it's best to not tackle the rate limiting at all. It's best (and easier of course) to let the users deal with it, especially since there's different ways to deal with it + there's the possibility some users are not even rate limited to begin with (not sure if some premium plan changes this for example).

As for the todos:

- For the domain test: I guess it's unfortunate if we can't try those out without an actual domain, we could mention this in the module docs perhaps. We technically could mock the responses by copying the ones they have in their examples but not sure there's much point tbh
- The `SendEmail::new` seems a bit over engineered but fine at the same time and I can't come up with something better. I mean it works so good enough haha
- I'm guessing the third todo was the rate limit note i left. I think it is fine for now since using 1 thread works. I might look into what you can do with a custom test harness or something in the future but not something I'll do anytime soon.

martsokha commented:
- Yeah, you can request an additional rate limit; that's precisely why I've added
the `RESEND_RATE_LIMIT` env variable.
The `tower::limit::RateLimit` middleware is the obvious choice for an
asynchronous client since `reqwest::Client` implements `tower::Services`.
Another viable option would be a custom
`tower::Layer::<http::Request, Response = http::Response>` as per `ietf` spec.
However, neither of these options is suitable for a blocking client.
- Same-same.

AntoniosBarotsis commented:
- I'm a bit skeptical with this, are there any points where we are *not* using references? I feel like even though references are less annoying for users, I can't think of a case where you would want to use the same request object more than once for instance (so I feel like forcing owned values in the parameters might prevent some bugs where people forget to pass the object at all). Another thing that might suck if we use references is if for whatever reason we need to switch to owned values later on which would be a breaking change. I think it is best we ignore the lint (on each instance of the warning though, not globally)
- For the const functions, from what I remember there's no penalty from making them const so why not I guess if non-const code can still call them then why not.
- I'll do these two myself as well at some point 
- This would be neat! I could ask on their Discord or something after this is merged

martsokha commented:
- As of now, we take `*Id`s by reference (as `Ids` are reusable) and request types by value, e.g. 
```rust
pub async fn create(&self, audience: &AudienceId, contact: ContactData) -> Result<ContactId> { ... }
```
- Alright, added `const`s.

Btw, what's the reason for a custom `config.toml`? Why not clippy with a pedantic flag? Not exactly the same, tho.

AntoniosBarotsis commented:
Yeah ig ids are fine as references and we can keep everything else as owned values. Any other thoughts on this?

I've mostly kept the config from back when I was still very new to the language, just out of habit. I should re-evaluate it at some point cause I bet it is annoying to deal with 😅. It is not just pedantic cause i like my editor spamming me with warnings

AntoniosBarotsis commented:
I'll take a look at the readme and the changelog but other than that everything is (finally 🎉) done right?

martsokha commented:
Yeah, I think so. I'll double-check in the evening.

AntoniosBarotsis reviewed (approved):
LGTM!
